README.md
[![Build Status](https://travis-ci.org/hmsuorg/icestream.svg?branch=master)](https://travis-ci.org/hmsuorg/icestream)
[![PyPI](https://img.shields.io/pypi/pyversions/Django.svg)](https://travis-ci.org/hmsuorg/icestream)

### HMSU Online Radio - Icestream

`icestream.py` is `CLI` tool, that play a role of wrapper, around gst-launch,
used for real time audio streaming, during HMSU Radio shows.

The tool requires `gstream` so you have to ensure it is installed.

#### Installation

First let's clone the project:

    $ git clone https://github.com/hmsuorg/icestream

Next let's install it:

    $ cd icestream
    $ pip install .

#### Usage

For help type:

    $ icestream.py --help

Example usage:

    $ icestream.py --password YOURPASSWORD --desc "HMSU Radio Show" --ip YOUR_ICECAST_SERVER_IP \
        --port 8888

the above command will create `lamemp3enc` and will stream the data to `YOUR_ICECAST_SERVE_IP` on port `8888`.

the default bitrate is `128kbps` if you wish to change the value use the `--bitrate` option.
The valid values are: `96`, `128`, `192`, `320` `kbps`.

#### Testing

    $ python setyp.py test

or

    $ python -m unittest discover

for code coverage:

    $ pip install coverage
    $ coverage run --source=. -m unittest
    $ coverage report -m

#### Contributing

For new a feature we using a `git flow` workflow, so please follow it. When you are ready push your feature.
Then will be merged into a `develop` branch.

Author: **Dimitar Dimitrov** <targy@hmsu.org>


CHANGELOG.md
Changelog
=========


0.4.0 (2018-04-10)
------------
- Merge branch 'release/0.4.0' Version 0.4.0 comes with unittests,
  reconnection functionality, the click options was improved, new
  defaults also was introduced. [Hakini]
- Bump version to 0.4.0. [Hakini]
- Docs, refactoring of an execute method, add gst as click option,
  rewrite the reconnect functionality. [Hakini]
- Refactoring and more unittests. [Hakini]
- Remove shutil.which, add --gst click option, shlex.split now is done
  in execute method not in cmd. [Hakini]
- Improve logging, separating warning log from info. [Hakini]
- Improve logging. [Hakini]
- Adding of reconnect strategy and error handler. [Hakini]
- Merge pull request #17 from hmsuorg/develop. [Di Mita Hakini]

  Fixes around travis-ci
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Remove py34 and py35. [Hakini]


0.3.7 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.7' Remove py34 and py35. [Hakini]
- Bump version to 0.3.7. [Hakini]
- Remove python3.4 and python3.5. [Hakini]
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Adding of status badges and more
  travis-ci tests. [Hakini]
- Merge branch 'master' into develop sync with github. [Hakini]


0.3.6 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.6' Adding of status badges and more travis-
  ci tests. [Hakini]
- Bump version to 0.3.6. [Hakini]
- Add badges and extend travis-ci and tox for py34 and py35. [Hakini]
- Merge pull request #16 from hmsuorg/develop. [Di Mita Hakini]

  Version 0.3.5 comes after many travis-ci fixes
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Remove coveralls from tox.ini.
  [Hakini]


0.3.5 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.5' Remove coveralls from tox.ini. [Hakini]
- Bump version to 0.3.5. [Hakini]
- Remove coveralls from tox. [Hakini]
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Fixes in use cases when gst is not
  installed. [Hakini]


0.3.4 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.4' Fixes in use cases when gst is not
  installed. [Hakini]
- Bump version to 0.3.4. [Hakini]
- Fixes around use cases when gst is not installed. [Hakini]
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Fix gst and tests to pass travis-
  ci. [Hakini]


0.3.3 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.3' Fix gst and tests to pass travis-ci.
  [Hakini]
- Bump version to 0.3.3. [Hakini]
- In case gst is missing we fallback to default path. [Hakini]
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Remove VERSION file in favor of
  setup.py. [Hakini]


0.3.2 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.2' Remove VERSION file in favor of setup.py.
  [Hakini]
- Remove VERSION file, setup.py will keep the version number. [Hakini]
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Fix VERSION path, because of
  travis-ci error. [Hakini]


0.3.1 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.1' Fix VERSION path, because of travis-ci
  error. [Hakini]
- Bump version to 0.3.1. [Hakini]
- Fix path of VERSION file, travis-ci fail with it. [Hakini]
- Merge branch 'master' into develop Github sync. [Hakini]
- Merge pull request #15 from hmsuorg/develop. [Di Mita Hakini]

  Improved README.md adding of .travis.yml and tox.ini files for CI
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Improved README.md adding of
  .travis.yml and tox.ini files for CI. [Hakini]


0.3.0 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.3.0' Improved README.md adding of .travis.yml
  and tox.ini files for CI. [Hakini]
- Bump version to 0.3.0 and improved README.md. [Hakini]
- Add .travis.yml and tox.ini files fix setup.py. [Hakini]
- Merge branch 'master' into develop Merge after git dev to master, pull
  request. [Hakini]
- Merge pull request #14 from hmsuorg/develop. [Di Mita Hakini]

  Sync via github Develop
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Fix setup.py and some other typos.
  [Hakini]


0.2.1 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.2.1' Fix setup.py and some other typos.
  [Hakini]
- Bump version to 0.2.1. [Hakini]
- Change bitrage from 192 to 128 in click defaults. [Hakini]
- Fix setup.py and update readme. [Hakini]
- Merge branch 'master' into develop. [Hakini]
- Merge pull request #13 from hmsuorg/develop. [Di Mita Hakini]

  Develop
- Hook: merge - update changelog. [Hakini]
- Merge branch 'master' into develop Version 0.2.0 comes with unittests
  and improved options. The options `--encoder` and `--save-to` was
  removed. Now they are supported by API by default. The click defaults
  also was improved. [Hakini]
- Merge branch 'master' into develop. [Hakini]


0.2.0 (2018-04-06)
------------------
- Hook: update chnagelog. [Hakini]
- Merge branch 'release/0.2.0' Version 0.2.0 comes with unittests and
  improved options. The options `--encoder` and `--save-to` was removed.
  Now they are supported by API by default. The click defaults also was
  improved. [Hakini]
- Bump version to 0.2.0. [Hakini]
- Write moreunittests and API improvments. [Hakini]
- Write unittests. [Hakini]
- Merge branch 'feature/testing' into develop remove encoder and save_to
  as click options, now they are done by default. [Hakini]
- Change click option defaults, remove save_to as options, now default
  operation. [Hakini]
- Lamemp3enc encoder is no longer a click option. [Hakini]

test_icestream.py
import unittest
from unittest.mock import MagicMock
import shutil
import shlex
import icestream

class TestIceStreamClick(unittest.TestCase):

    def setUp(self):
        self.__gst = 'gst-launch-1.0'

        self.__params = {
            'source': 'alsasrc',
            'bitrate': 128,
            'ip': 'radio.hmsu.org',
            'port': 8000,
            'password': None,
            'genre': 'drum and bass',
            'streamname': 'HMSU Online Radio',
            'desc': 'The Colours Of Drum and Bass',
            'gst': 'gst-launch-1.0'
        }


    def test_icestream_without_args(self):
        result = icestream.IceStream(**self.__params).cmd()
        result = shlex.split(result)
        self.assertEqual(result, shlex.split('{} alsasrc ! queue ! audioconvert ! lamemp3enc bitrate=128 ! tee name=t t. ! queue ! shout2send ip=radio.hmsu.org port=8000 password="None" mount=/bass -t genre="drum and bass" streamname="HMSU Online Radio" description="The Colours Of Drum and Bass" t. ! queue ! filesink location=HMSU_Online_Radio_The_Colours_Of_Drum_and_Bass.mp3'.format(self.__gst)))

    def test_icestream_password_args(self):
        self.__params['password'] = 'yourpassword'
        result = icestream.IceStream(**self.__params).cmd()
        result = shlex.split(result)
        self.assertEqual(result, shlex.split('{} alsasrc ! queue ! audioconvert ! lamemp3enc bitrate=128 ! tee name=t t. ! queue ! shout2send ip=radio.hmsu.org port=8000 password="yourpassword" mount=/bass -t genre="drum and bass" streamname="HMSU Online Radio" description="The Colours Of Drum and Bass" t. ! queue ! filesink location=HMSU_Online_Radio_The_Colours_Of_Drum_and_Bass.mp3'.format(self.__gst)))

    def test_icestream_password_bitrate_wrong_args(self):
        self.__params['password'] = 'yourpassword'
        self.__params['bitrate'] = 222
        with self.assertRaises(Exception) as context:
            icestream.IceStream(**self.__params).cmd()
        self.assertTrue('Allowd bitrates are' in str(context.exception))

    def test_icestream_password_bitrate_args(self):
        self.__params['password'] = 'yourpassword'
        self.__params['bitrate'] = 192
        result = icestream.IceStream(**self.__params).cmd()
        result = shlex.split(result)
        self.assertEqual(result, shlex.split('{} alsasrc ! queue ! audioconvert ! lamemp3enc bitrate=192 ! tee name=t t. ! queue ! shout2send ip=radio.hmsu.org port=8000 password="yourpassword" mount=/bass -t genre="drum and bass" streamname="HMSU Online Radio" description="The Colours Of Drum and Bass" t. ! queue ! filesink location=HMSU_Online_Radio_The_Colours_Of_Drum_and_Bass.mp3'.format(self.__gst)))

    def test_icestream_password_bitrate_ip_port_args(self):

        self.__params['password'] = 'yourpassword'
        self.__params['bitrate'] = 192
        self.__params['ip'] = 'radio2.hmsu.org'
        self.__params['port'] = 9999

        result = icestream.IceStream(**self.__params).cmd()
        result = shlex.split(result)
        self.assertEqual(result, shlex.split('{} alsasrc ! queue ! audioconvert ! lamemp3enc bitrate=192 ! tee name=t t. ! queue ! shout2send ip=radio2.hmsu.org port=9999 password="yourpassword" mount=/bass -t genre="drum and bass" streamname="HMSU Online Radio" description="The Colours Of Drum and Bass" t. ! queue ! filesink location=HMSU_Online_Radio_The_Colours_Of_Drum_and_Bass.mp3'.format(self.__gst)))

    def test_icestream_password_bitrate_ip_port_genre_args(self):

        self.__params['password'] = 'yourpassword'
        self.__params['bitrate'] = 192
        self.__params['ip'] = 'radio2.hmsu.org'
        self.__params['port'] = 9999
        self.__params['genre'] = 'dnb'
        self.__params['desc'] = 'dnb desc'
        self.__params['streamname'] = 'dnb stream'

        result = icestream.IceStream(**self.__params).cmd()
        result = shlex.split(result)
        self.assertEqual(result, shlex.split('{} alsasrc ! queue ! audioconvert ! lamemp3enc bitrate=192 ! tee name=t t. ! queue ! shout2send ip=radio2.hmsu.org port=9999 password="yourpassword" mount=/bass -t genre="dnb" streamname="dnb stream" description="dnb desc" t. ! queue ! filesink location=dnb_stream_dnb_desc.mp3'.format(self.__gst)))

    def test_icestream_gst(self):
        result = icestream.IceStream(**self.__params)
        self.assertEqual('gst-launch-1.0', result.gst)

    def test_icestream_execute_with_error(self):
        result = icestream.IceStream(**self.__params)
        res = result.execute('head -n1 README1.md')
        self.assertFalse(res)

    def test_icestream_execute_not_found(self):
        result = icestream.IceStream(**self.__params)

        ices = icestream.IceStream(**self.__params)
        result = ices.execute('head1')
        self.assertFalse(result)


icestream.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
This scripts provide a gst-launch wrapper, used for streaming HMSU radio shows.
Author: targy@hmsu.org
"""
import time
import datetime
import subprocess
import shlex
import click


class IceStream:
    """IceStream"""

    def __init__(self, **kwargs):
        """__init__"""

        self.__bitrates = [96, 128, 192, 320]
        self.gst = kwargs.get('gst')
        self.__params = kwargs

    def cmd(self):
        """cmd method will generate a gst-launch command,
        that will make stream between gst-launch and icecast stream"""

        cmd = '{gst} {source} ! queue ! audioconvert ! lamemp3enc bitrate={bitrate} '

        if self.__params['bitrate'] not in self.__bitrates:
            raise Exception('Allowd bitrates are: {}'.format(self.__bitrates))

        # save_to
        cmd += '! tee name=t t. ! queue '

        cmd += '! shout2send ip={ip} port={port} password="{password}" mount=/bass -t genre="{genre}" '
        cmd += 'streamname="{streamname}" description="{desc}" '

        # save_to
        self.__params['save_to'] = '{}_{}.mp3'.format(
            self.__params.get('streamname'), self.__params.get('desc')).replace(' ', '_')
        cmd += 't. ! queue ! filesink location={save_to}'

        return cmd.format(**self.__params)

    def execute(self, cmd: str) -> bool:
        """execute a shell command. it's designed to run gst-launch,
        but, could be anything else.help

        :param cmd: a shell command
        :type cmd: str
        :return: bool
        :rtype: bool
        """

        cmd = shlex.split(cmd)

        try:

            with subprocess.Popen(
                cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=1, universal_newlines=True) as p:

                click.secho('+++ Connecting to server: {ip} on port: {port} ...'.format(**self.__params), fg='cyan')

                for line in p.stdout:
                    line = line.replace('\n', '')

                    if line.startswith('WARNING'):
                        click.secho('!!! ' + line, fg='yellow')

                    else:
                        click.secho('*** INFO ' + line, fg='green')

                error = p.stderr.readlines()

        except FileNotFoundError as error:
            # we got this exception in case when gst-launch isn't installed
            click.secho(str(error), fg='red')
            return False

        else:

            if p.returncode != 0:
                error_info = '--- ERROR: {}: {} on port: {}'.format(
                    error[0].split(':')[-1].strip(), self.__params.get('ip'), self.__params.get('port'))

                error_add = error[-1].split(':')[-1].strip().split('=')[-1]

                click.secho('{} -- {}, {}'.format(datetime.datetime.now(), error_info, error_add), fg='red')

                return False

            return True

@click.command()
@click.option('--gst', default='gst-launch-1.0', help='gst-launch executable')
@click.option('--source', default='alsasrc', help='gst-launch source, default is alsasrc')
@click.option('--bitrate', default=128, help='gst-launch encoder bitrate, default is 128kbps')
@click.option('--ip', default='radio.hmsu.org',
              help='icecast ip or hostname default is radio.hmsu.org')
@click.option('--port', default=8000, help='icecast port default is 8000')
@click.option('--password', help='icecast password')
@click.option('--genre', default='drum and bass',
              help='icecast metadata - stream genre, default is dnb')
@click.option('--streamname', default='HMSU Online Radio',
              help='icecast metadata - stream name, defautl is HMSU Radio')
@click.option('--desc', default='The Colours Of Drum and Bass', help='icecast metadata - stream description aka tcodnb')
def main(**kwargs):
    """icestream - gst-launch to icecast radio tool"""

    ices = IceStream(**kwargs)
    while ices.execute(ices.cmd()) is False:
        time.sleep(3)

if __name__ == "__main__":
    main()
